_posts/2023-08-07-biquery-partitioning.md

```diff
@@ -0,0 +1,166 @@
+---
+title: "BigQuery partitioning - by time-unit column or by ingestion time"
+date: 2023-08-07
+tags: [GCP, BigQuery]
+header:
+    overlay_image: /assets/images/2022-03-24-gcp-dataproc-spark-tuning/nana-smirnova-IEiAmhXehwE-unsplash.webp
+    caption: "[Unsplash](https://unsplash.com/@nanasmirnova)"
+---
+
+What's the best way to partition time-series data in BigQuery?
+By time-unit column or by ingestion time? Daily or hourly?
+It depends, keep reading to learn trade-offs, pitfalls, and other traps.
+
+## Time-series or aggregates
+
+For time-series with data points with minute, second or millisecond precision use time-unit column partitioning.
+You will get convenient querying like:
+
+```sql
+SELECT ts, temperate, pressure
+FROM weather_stations
+WHERE
+    ts BETWEEN TIMESTAMP("2008-12-25 15:30:00")
+        AND TIMESTAMP("2008-12-25 15:35:00")
+```
+
+With ingestion time partitioning, you have to specify extra predicate for partition pruning:
+
+```sql
+SELECT ts, temperate, pressure
+FROM weather_stations
+WHERE
+    ts BETWEEN TIMESTAMP("2008-12-25 15:30:00")
+        AND TIMESTAMP("2008-12-25 15:35:00")
+    AND _PARTITIONTIME = DATE("2008-12-15")
+```
+
+For aggregates, when you typically don't need an exact point in time, use ingestion time partitioning.
+You don't need to specify the time column explicitly in a table schema.
+
+```sql
+SELECT _PARTITIONTIME AS dt, page_views, unique_visitors
+FROM ecommerce_sessions
+WHERE
+    _PARTITIONTIME = DATE("2008-12-15")
+```
+
+## Retention
+
+In my projects, the majority of tables require at least 1--3 years of history.
+With a limit of 4000 partitions per BigQuery table, it requires at least daily partitioning.
+Tables with 3 years of retention use `3 * 365 = 1095` daily partitions, below limit.
+Tables with hourly partitions keep up to only `4000 / 24 = 166 days and 8 hours` of data.
+For tables with more than 10 years of history I would consider monthly partitioning.
```

Comment:
Myślę, że warto zasugerować też trzymanie w dziennych z retencją partycji na te powiedzmy 3 lata, a starsze dane jakimś mechanizmem przenosić na tańszy storage (np. GCS archive). Zależy nam w BQ na ograniczeniu podnoszonych danych (niezależnie od pricingu). Miesięczne partycje nie będą optymalne, jeśli use case związany z odczytem będzie potrzebował filtrowania pojedynczych dni.

Reply:
Oczywiście można też zrobić miesięczne i dodać klastrowanie po dniu (tak jak zasugerowałeś niżej) - nie mniej jednak zostawiam powyższy komentarz, bo jest wcześniej w tekście i jeśli ktoś nie doczyta do końca to wyjdzie z być może złymi wnioskami :) Może jakiś link do niższej sekcji i tego tricku z klastrowaniem?  
(Z tym trickiem jest tylko taki problem, że bezpośrednio w konsolce nie oszacujemy kosztów)

Reply:
słusznie, jak ktoś potrzebuje tyle historii to offload na GCS i external table wydaje się pierwszym wyborem, na odczepnego można kierować tu https://cloud.google.com/bigquery/docs/best-practices-storage#archival-data-cloud-storage

Reply:
@dkrynski - wywaliłem weekly partitioning at all, skoro nie stosuję (a offloading tak) to nie powinienem pisać bzdur